sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java

```diff
@@ -58,6 +58,44 @@
 
 /**
  * A {@link ChannelPool} implementation that enforces a maximum number of concurrent direct TCP Cosmos connections.
+ *
+ * RntbdClientChannelPool: Actors
+ * 	- acquire (RntbdServiceEndpoint): acquire a channel to use
+ * 	- release (RntbdServiceEndpoint): channel usage is complete and returning it back to pool
+ * 	- Channel.closeChannel() Future: Event handling notifying the channel termination to refresh bookkeeping
+ * 	- acquisitionTimeoutTimer: channel acquisition wait-out handler
+ * 	- monitoring (through RntbdServiceEndpoint): get monitoring metrics
+ *
+ * 	Behaviors/Expectations:
+ * 	    - Bounds: MAX_CHANNELS_PER_ENDPOINT * MAX_REQUESTS_ENDPOINT (NOT A GUARANTEE)
+ * 	    - NewChannel vs ReUseChannel:
+ * 	        - NewChannels are serially created (current state NOT by-design)
```

Comment:
I would argue this is by design - impossible to guarantee max TCP connections per endpoint without accepting that you might have to throw away newly created connections - establishing a connection is expensive - for both client and service. Usage scenarios don't need to be optimized for cold-start latency - CosmosDB prime time is for high load /throughput scenarios

Reply:
On Max TCP connections part: Optimistically can be on done through Semaphore (with tryAcquire). 

I do agree that creating all connection is unnecessary and also counter productive on resources (as you mentioned). With OOB 130 connections are neither in great shape there. This needs revisit.

Will update the text on NOT by-design part.